astropy/modeling/bounding_box.py

```diff
@@ -0,0 +1,573 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+
+"""
+This module is to contain an improved bounding box
+"""
+
+from collections import namedtuple
+from typing import TYPE_CHECKING, Dict
+
+from astropy.utils import isiterable
+
+import warnings
+import numpy as np
+
+if TYPE_CHECKING:
+    from .core import Model
+
+
+__all__ = ['Interval', 'BoundingBox']
+
+
+_BaseInterval = namedtuple('_BaseInterval', "lower upper")
+
+
+class Interval(_BaseInterval):
+    """
+    A single input's bounding box interval.
+
+    Parameters
+    ----------
+    lower : float
+        The lower bound of the interval
+
+    upper : float
+        The upper bound of the interval
+
+    Methods
+    -------
+    validate :
+        Contructs a valid interval
+
+    outside :
+        Determine which parts of an input array are outside the interval.
+
+    domain :
+        Contructs a discretization of the points inside the interval.
+    """
+
+    @staticmethod
+    def _validate_shape(interval):
+        """Validate the shape of an interval representation"""
+        MESSAGE = """An interval must be some sort of sequence of length 2"""
+
+        try:
+            shape = np.shape(interval)
+        except TypeError:
+            try:
+                # np.shape does not work with lists of Quantities
+                if len(interval) == 1:
+                    interval = interval[0]
+                shape = np.shape([b.to_value() for b in interval])
+            except (ValueError, TypeError, AttributeError):
+                raise ValueError(MESSAGE)
+
+        valid_shape = shape in ((2,), (1, 2), (2, 0))
+        if not valid_shape:
+            valid_shape = (len(shape) > 0) and (shape[0] == 2) and \
+                all(isinstance(b, np.ndarray) for b in interval)
+
+        if not isiterable(interval) or not valid_shape:
+            raise ValueError(MESSAGE)
+
+    @classmethod
+    def _validate_bounds(cls, lower, upper):
+        """Validate the bounds are reasonable and construct an interval from them."""
```

Comment:
This is to "support" model sets; however, model sets have major issues with bounding box as it is. See issue #11902.

This exists so that if a user requests the bounding box for a model set (`mdl.bounding_box`) which defines a bounding box via function, eg `Gaussian1D()` no validation errors will occur.

Note that without this, the following test was failing: https://github.com/astropy/astropy/blob/00b7498a152c1be67104803beac1df491302bd06/astropy/modeling/tests/test_compound.py#L524-L539